spyder/plugins/ipythonconsole/comms/kernelcomm.py

```diff
@@ -0,0 +1,107 @@
+from qtpy.QtCore import QEventLoop, QObject, QTimer, Signal
+
+from spyder_kernels.comms.commbase import CommBase
+from spyder_kernels.py3compat import TimeoutError
+
+logger = logging.getLogger(__name__)
+
+
+class KernelComm(CommBase, QObject):
+    """
+    Class with the necessary attributes and methods to handle
+    communications with a console.
+    """
+
+    sig_got_reply = Signal(str)
+    sig_debugging = Signal(bool)
+
+    def __init__(self):
+        super(KernelComm, self).__init__()
+
+        self.kernel_client = None
+        self._debugging = False
+        self._wait_list = {}
+
+        self.register_call_handler(
+            'set_debug_state', self._handle_debug_state)
+
+    def set_kernel_client(self, kernel_client):
+        """Register new kernel client and open comm."""
+        self._register_comm(
+            kernel_client.comm_manager.new_comm(self._comm_name))
+        self.kernel_client = kernel_client
+
+    def remote_call(self, interrupt=False, blocking=False, comm_id=None):
+        """Get a handler for remote calls."""
+        return super(KernelComm, self).remote_call(
+            interrupt=interrupt, blocking=blocking, comm_id=comm_id)
+
+    # ---- Private -----
+    def _get_call_return_value(self, call_dict):
+        """
+        Interupt the kernel if needed.
+        """
+        settings = call_dict['settings']
+        if 'interrupt' in settings and settings['interrupt']:
+            self._signal_update_kernel()
+
+        return super(KernelComm, self)._get_call_return_value(call_dict)
+
+    def _signal_update_kernel(self):
+        """Interrupt the kernel to give a chance to read other messages."""
+        self._pdb_update()
+
+    def _pdb_update(self):
+        """
+        Update by sending an input to pdb.
+        """
+        if self._debugging and self.kernel_client:
+            cmd = (u"!get_ipython().kernel.frontend_comm" +
+                   ".remote_call(blocking=True).pong()")
+            self.kernel_client.input(cmd)
+
+    def _handle_debug_state(self, is_debugging):
+        """Update the debug state."""
+        self._debugging = is_debugging
+        self.sig_debugging.emit(is_debugging)
+
+    def _wait_reply(self, call_id, call_name, timeout):
+        """Wait for the other side reply."""
+        if call_id in self._call_reply_dict:
+            return
+        self._wait_list[call_id] = call_name
+
+        # Create event loop to wait with
+        wait_loop = QEventLoop()
+        self.sig_got_reply.connect(wait_loop.quit)
+        wait_timeout = QTimer()
+        wait_timeout.setSingleShot(True)
+        wait_timeout.timeout.connect(wait_loop.quit)
+
+        # Wait until the kernel returns the value
+        wait_timeout.start(timeout * 1000)
+        while len(self._wait_list) > 0:
+            if not wait_timeout.isActive():
+                raise TimeoutError(
+                    "Timeout while waiting for {}".format(
+                        self._wait_list))
```

Comment:
I think this error should be shown in a QMessageBox instead of raising it. That's because a raise would cause our error report dialog to be displayed, which would make users to post this to Github. And (it seems to me) this is informative and not a cause of concern.

Reply:
Maybe it would be better to handle that where the blocking call is made? Or alternaltively to disable blocking calls from the frontend and only use callbacks?

Reply:
Actually it makes a lot of sense to add to option to have a callback on the remote calls. The blocking option should only be used when we need to block the thread.

Reply:
> Actually it makes a lot of sense to add to option to have a callback on the remote calls

Ok, but that still doesn't solve my comment here. Please show here a QMessageBox instead of raising an error, so users don't post this error on Github.

Reply:
So the problem here is that the calling function expects a reply, so the QMessageBox should be done there.

Reply:
If I just show a QMessageBox at this point and return None, any function that calls the kernel in a blocking manner might get a None instead of the expected value? Or we expect any functions that makes a blocking call to handle a timeout?

Reply:
Is it ok for the following functions to return None in case of timeout?
```
is_defined
get_doc
get_source
get_value
load_data
save_namespace
```

Reply:
> Is it ok for the following functions to return None in case of timeout?

It was like that before (you can check that in the code you removed in this PR). And we handled things that way because most of the methods you mentioned above were not supported while the kernel was on debugging mode (are they working while debugging now?)

The only exception was `get_value`. In that case, if we were not able to retrieve a value, we sent the traceback generated in the kernel when trying to do that and showed it in another QMessageBox.

Reply:
ok I'll add a try: to these functions to catch the timeout and return None